_posts/2020-06-01-rails-fejlesztes-dockerben.md

```diff
@@ -0,0 +1,233 @@
+- Több alkalmazás párhuzamos fejlesztése adott függőség (pl.: Adatbáziskezelő alakalmazás) eltérő verzióinak használatával könnyen megoldható, mivel könnyen lehet párhuzamosan használni több konténerhálózatot.
+
+## Konténerek elkészítése
+
+A JetBrains csapata elkészitett egy [példa projektet][5], a konténerek elkészítéséhez azt vesszük alapul. Elösször szükségünk van egy konténerre amiben, a Rails keretrendszert tudjuk futtatni. Docker Hub-on nem találtam megfelelő image-et ehez, ezért az alábbi **Dockerfilet**-ban érdemes leirni az image-et:
+
+```Dockerfile
+FROM ruby:2.6.3
+RUN curl -sL https://dl.yarnpkg.com/debian/pubkey.gpg | apt-key add -
+RUN echo "deb https://dl.yarnpkg.com/debian/ stable main" | tee /etc/apt/sources.list.d/yarn.list
+RUN apt-get update -qq && apt-get install -y build-essential libpq-dev nodejs postgresql-client yarn
+RUN mkdir /test_app
+WORKDIR /test_app
+COPY Gemfile /test_app/Gemfile
+COPY Gemfile.lock /test_app/Gemfile.lock
+COPY package.json /test_app/package.json
+COPY yarn.lock /test_app/yarn.lock
+RUN gem install bundler -v '2.0.2'
+RUN bundle install
+RUN yarn install --check-files
+COPY . /test_app
+
+EXPOSE 3000
+```
+
+A fenti image a Ruby 2.6.3-as verzióját tartalmazza, egy Debian rendszeren. Telepítjük még a Yarn és NodeJS csomagokat, amik kezelik a projectben lévő JavaScript fájl-okat. Ezután létrehozzuk a test_app mappát és beállítjuk a konténer munkamappájának. Az ezt követő rész felelős az alkalmazás függőségeit képező Ruby és JavaScript könyvtárak letöltéséért. Az utolsó két sor bemásolja a project mappáját a konténerbe, majd a konténer 3000-es portját elérthezővé teszi a gazdagép számára.
+
+A Dockerfile elkészítése után elkészíthetjük a **docker-compose.yml** fájl-t, ami a több konténerből álló alkalmazásunkat írja le.
+
+
+```docker
+version: '3'
+services:
+  db:
+    image: postgres
+    volumes:
+      - ./tmp/db:/var/lib/postgresql/data
+    environment:
+      POSTGRES_HOST_AUTH_METHOD: trust
+    ports:
+      - "5432:5432"
+  web:
+    build: .
+    # command: bundle exec rails s -p 3000 -b '0.0.0.0'
+    command: tail -f /dev/null
+    volumes:
+      - .:/test_app
+      - /test_app/node_modules
+    ports:
+      - "3000:3000"
+      # Ports required for debugging
+      - "1234:1234"
+      - "26166:26168"
+    depends_on:
+      - db
+  ```
+
+A fájl első részében létrehozzuk a **db** szolgáltást, ami a postgres image alapú konténer lesz. Az adatbázisban tárolt adatok mappáját összekötjük a host egyik mappájával, ezzel perzisztensé téve az adatbázis adatait. Az egyszerűség kedvéért jelszót nem állítunk be. Majd a host 5432-es portját és a konténer 5432-es portját összekötjük.
+
+A második szolgáltatásban fog futni a rails web szerver, ezért ezt **web** néven hozzuk létrehozni. Két lehetséges parancsot is felvettem. Az első elinditja a szervert. A második csak futó állapotban tartja a konténer, hogy fejlsztés közben csatlakozni lehessen hozzá.
```

Comment:
A második szolgáltatásban fog futni a rails web szerver, ezért ezt **web** néven fogjuk létrehozni. Két lehetséges parancsot is felvettem. Az első elindítja a szervert. A második csak futó állapotban tartja a konténert, hogy fejlesztés közben csatlakozni lehessen hozzá.